File: Python/MedianofTwoSortedArrays.py
import math
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        if(len(nums1) > len(nums2)): return self.findMedianSortedArrays(nums2, nums1)

        m = len(nums1)
        n = len(nums2)

        left = 0
        right = m

        while left <= right:

            cut1 = math.floor((left+right)/2)
            cut2 = (math.floor((m+n+1)/2)) - cut1

            Aleft = float('-inf') if cut1 == 0 else nums1[cut1-1]
            Aright = float("inf") if cut1 == m else nums1[cut1]

            Bleft = float('-inf') if cut2 == 0 else nums2[cut2-1]
            Bright = float('inf') if cut2 == n else nums2[cut2]

            if Aleft <= Bright and Bleft <= Aright:
                if (m+n) % 2 == 1: 
                    return max(Aleft, Bleft)
                return (max(Aleft,Bleft)+ min(Aright, Bright))/2

            if Aleft > Bright:
                right = cut1-1
            else: left = cut1+1


solution = Solution()

File: JavaScript/MedianofTwoSortedArrays.js
const findMedianSortedArrays = function (nums1, nums2) {
  if (nums1.length > nums2.length) {
    return findMedianSortedArrays(nums2, nums1);
  }

  const m = nums1.length;
  const n = nums2.length;

  let left = 0;
  let right = m;

  while (left <= right) {
    const cut1 = Math.floor((left + right) / 2);

    const cut2 = Math.floor((m + n + 1) / 2) - cut1;

    const Aleft = cut1 === 0 ? -Infinity : nums1[cut1 - 1];
    const Aright = cut1 === m ? Infinity : nums1[cut1];

    const Bleft = cut2 === 0 ? -Infinity : nums2[cut2 - 1];
    const Bright = cut2 === n ? Infinity : nums2[cut2];

    if (Aleft <= Bright && Bleft <= Aright) {
      if ((m + n) % 2 === 1) return Math.max(Aleft, Bleft);

      return (Math.max(Aleft, Bleft) + Math.min(Aright, Bright)) / 2;
    }

    if (Aleft > Bright) {
      right = cut1 - 1;
    } else {
      left = cut1 + 1;
    }
  }
};

findMedianSortedArrays([1, 2], [3, 4]);
